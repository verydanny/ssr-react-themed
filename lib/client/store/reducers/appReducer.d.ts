export interface AppState {
    app_loaded: boolean;
}
export declare const INITIAL_STATE: Readonly<{
    app_loaded: boolean;
}>;
declare const _default: (state: AppState | undefined, action: {
    type: string;
    payload?: any;
}) => AppState;
export default _default;
