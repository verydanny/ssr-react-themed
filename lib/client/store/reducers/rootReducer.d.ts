declare const rootReducer: import("redux").Reducer<{
    app: import("./appReducer").AppState;
    router: import("react-router-redux").RouterState;
}, import("redux").AnyAction>;
export default rootReducer;
