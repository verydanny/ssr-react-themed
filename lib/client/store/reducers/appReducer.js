"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const app_1 = require("../actions/app");
exports.INITIAL_STATE = Object.freeze({
    app_loaded: false
});
exports.default = (state = exports.INITIAL_STATE, action) => {
    const { type, payload = {} } = action;
    switch (type) {
        case app_1.ActionTypes.APP_LOADED: {
            return {
                ...state,
                app_loaded: true
            };
        }
    }
    return state;
};
