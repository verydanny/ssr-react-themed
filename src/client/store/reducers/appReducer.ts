import { ActionT } from '../types'
import { ActionTypes } from '../actions/app'

export interface AppState {
  app_loaded: boolean
}

export const INITIAL_STATE = Object.freeze({
  app_loaded: false
})

export default (state: AppState = INITIAL_STATE, action: ActionT) => {
  const { type, payload = {} } = action

  switch (type) {
    case ActionTypes.APP_LOADED: {
      return {
        ...state,
        app_loaded: true
      }
    }
  }

  return state
}