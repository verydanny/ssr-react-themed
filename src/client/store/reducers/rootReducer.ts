import { combineReducers } from 'redux'
import { routerReducer as router } from 'react-router-redux'

import app from './appReducer'

const rootReducer = combineReducers({
  app,
  router,
})

export default rootReducer