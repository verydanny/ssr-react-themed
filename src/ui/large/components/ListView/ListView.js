import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { themed } from 'react-themed'
import throttle from 'lodash/throttle'
import { LIST_SKELETON, NEXT_PAGE, PREV_PAGE } from 'app/store/types'
import { forceCheck } from 'react-lazyload'
import isElementInViewport from 'lib/utils/isElementInViewport'
import Footer from 'ui/shared/components/Footer'
import Pagination from 'ui/large/components/Pagination'
import BreadCrumbs from 'ui/shared/components/BreadCrumbs'
import pick from 'lodash/fp/pick'
import {
  LEAD_MODAL_PROPS,
} from 'ui/large/components/LeadModal/const'
import {
  LEAD_MODAL_ID,
  MORE_FILTERS_MODAL_ID,
} from 'ui/shared/config/modal-const'
import autobind from 'autobind-decorator'
import Filters from 'ui/large/components/Filters'
import SeoContent from 'ui/large/components/SeoContent'
import { DfpAdSlot } from 'ui/shared/components/DfpAdSlot'
import { LARGE_SRP_FOOTER_AD_KEY } from 'config/dfpAds'

import isEqual from 'lodash/isEqual'
import Cards from './ListingCard/Cards'
import LoadingCards from './ListingCard/LoadingCards'
import ZeroResultsCard from './ListingCard/ZeroResultsCard'
import ListHeader from './ListHeader'

@themed('*')
export default class ListView extends Component {

  static propTypes = {
    theme: PropTypes.object,
    listings: PropTypes.array,
    mapZoomOutOnEmpty: PropTypes.func,
    openModal: PropTypes.func,
    total: PropTypes.number,
    selectedListingId: PropTypes.string,
    selectedListingIsClicked: PropTypes.bool,
    selectedListingShouldScroll: PropTypes.bool,
    selectListingFromMapPin: PropTypes.func,
    clearListingData: PropTypes.func,
    clearListingScroll: PropTypes.func,
    loading: PropTypes.any,
    loadingOptions: PropTypes.object,
    favorites: PropTypes.object,
    toggleFavorite: PropTypes.func,
  }

  static defaultProps = {
    theme: {},
  }

  constructor(props) {
    super(props)
    this.handleScrollEvent = throttle(this.handleScrollEvent, 250)
  }

  componentDidMount() {
    this.scrollToSelectedListingIfNeeded()
  }

  componentWillReceiveProps(props) {
    if (!isEqual(props.listings, this.props.listings) ||
      (!this.props.loading && props.loading &&
        (props.loadingOptions.source === NEXT_PAGE || props.loadingOptions.source === PREV_PAGE))
    ) {
      this.scrollToTop()
    }
  }

  componentDidUpdate() {
    this.scrollToSelectedListingIfNeeded()
  }

  scrollToSelectedListingIfNeeded() {
    const {
      clearListingScroll,
      selectedListingShouldScroll,
    } = this.props

    if (selectedListingShouldScroll && clearListingScroll) {
      this.scrollToSelectedListing()
      clearListingScroll()
    }
  }

  scrollToSelectedListing() {
    // Find the active card in the list (it will be marked with an attribute)
    const card = this.listDiv.querySelector('[data-active-card]')

    if (card && !isElementInViewport(card)) {
      card.scrollIntoView()
    }
  }

  handleScrollEvent() {
    // Scroll events on elements with overflow:auto do not bubble up;
    // therefore the lazy loader will not know when content scrolls into view.
    // To ensure that content within the list can be lazy loaded, we
    // must manually tell the lazy loader to check if content is visible,
    // whenever the user scrolls the list.
    forceCheck()
  }

  @autobind
  handleCardMouseEnter(event, listingId) {
    const { selectListingFromMapPin } = this.props

    if (selectListingFromMapPin && listingId) {
      // Select a pin on the map, override any pin that was already clicked
      selectListingFromMapPin(listingId, {
        clicked: true,
        trackClick: false,
      })
    }
  }

  @autobind
  handleCardMouseLeave() {
    const { clearListingData } = this.props

    if (clearListingData) {
      // Clear the selected pin on the map
      clearListingData()
    }
  }

  @autobind
  showLeadModal(listing) {
    const { openModal } = this.props
    const listingProps = pick(LEAD_MODAL_PROPS)(listing)

    if (openModal) openModal(LEAD_MODAL_ID, listingProps)
  }

  scrollToTop() {
    this.cardsContainer.scrollTop = 0
  }

  @autobind
  showMoreFiltersModal() {
    const { openModal } = this.props

    if (openModal) openModal(MORE_FILTERS_MODAL_ID)
  }

  renderListContent() {
    const {
      loading,
      total,
      loadingOptions,
      mapZoomOutOnEmpty,
      selectedListingId,
      selectedListingIsClicked,
      listings,
      favorites,
      toggleFavorite,
    } = this.props

    if (!loading && total === 0) {
      return (
        <ZeroResultsCard
          onClickZoomOut={mapZoomOutOnEmpty}
          onClickOpenModal={this.showMoreFiltersModal}
        />
      )
    }

    let cards

    if (loading &&
      loadingOptions &&
      loadingOptions.skeleton !== LIST_SKELETON) {
      cards = LoadingCards({ listings })
    } else {
      cards = Cards({
        listings,
        selectedListingId,
        selectedListingIsClicked,
        favorites,
        toggleFavorite,
        openModal: this.showLeadModal,
        onCardMouseEnter: this.handleCardMouseEnter,
        onCardMouseLeave: this.handleCardMouseLeave,
      })
    }
    return (
      <div>
        <ListHeader
          total={total}
          loading={loading}
          loadingOptions={loadingOptions}
        />
        <div data-tag_section="property_listings">
          {cards}
        </div>
        <Pagination />
        <SeoContent />
      </div>
    )
  }

  render() {
    const {
      theme,
    } = this.props

    return (
      <div
        className={theme.ListView}
        ref={list => { this.listDiv = list }}
        onScroll={this.handleScrollEvent}
      >
        <BreadCrumbs />
        <Filters
          onOpen={this.onOpen}
          theme={theme}
          openModal={this.showMoreFiltersModal}
        />
        <div
          className={theme.List_Content}
          data-tid="list-view"
          ref={div => { this.cardsContainer = div }}
        >
          {this.renderListContent()}
          <div className={theme.List_Footer}>
            <Footer />
            <DfpAdSlot
              adKey={LARGE_SRP_FOOTER_AD_KEY}
              tid="srp-footer-ad"
            />
          </div>
        </div>
      </div>
    )
  }
}
