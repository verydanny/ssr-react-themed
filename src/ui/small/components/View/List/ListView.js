import React, { PureComponent } from 'react'
import BreadCrumbs from 'ui/shared/components/BreadCrumbs'
import autobind from 'autobind-decorator'
import PropTypes from 'prop-types'
import { themed } from 'react-themed-too'
import throttle from 'lodash/throttle'
import { forceCheck } from 'react-lazyload'
import { SMALL_LIST_VIEW_AD_KEY } from 'config/dfpAds'
import Footer from 'ui/shared/components/Footer'
import { Listings, AD_POSITION } from 'ui/small/components/ListingCard'
import SeoContent from 'ui/small/components/SeoContent'

@themed('List', { pure: true })

export default class ListView extends PureComponent {
  static propTypes = {
    loadedAds: PropTypes.array,
    title: PropTypes.string,
    city: PropTypes.string,
    count: PropTypes.number,
    currentView: PropTypes.string,
    hidden: PropTypes.bool,
    listings: PropTypes.array,
    loading: PropTypes.bool,
    loadingOptions: PropTypes.object,
    nextPageLink: PropTypes.string,
    openModal: PropTypes.func,
    pageNumber: PropTypes.number,
    pathname: PropTypes.string,
    previousPageLink: PropTypes.string,
    pushState: PropTypes.func,
    query: PropTypes.object,
    selectedListingId: PropTypes.string,
    selectListing: PropTypes.func,
    toggleView: PropTypes.func,
    total: PropTypes.number,
    locationSlug: PropTypes.string,
    searchType: PropTypes.string,
    propertyType: PropTypes.string,
    refinementCriteria: PropTypes.object,
    renderAd: PropTypes.func,
    theme: PropTypes.object,
    uri: PropTypes.string,
  }

  static defaultProps = {
    theme: {},
  }

  componentDidMount() {
    this.lazyCheck()

    const { selectedListingId } = this.props

    if (selectedListingId) this.scrollToListing(selectedListingId, true)

    // Ensure lazyCheck doesn't get called too often
    this.lazyCheck = throttle(this.lazyCheck, 500, { leading: false })

    // Call lazyCheck whenever the list is scrolled, so the lazy loader
    // can determine if any new cards are visible and load images
    this.listContainerElement.addEventListener('scroll', this.lazyCheck, true)
  }

  componentDidUpdate(prevProps) {
    const {
      selectedListingId,
      currentView,
      uri,
    } = this.props

    if (currentView === 'list') {
      this.lazyCheck()

      const switchFromMapToList = selectedListingId && prevProps.currentView === 'map'

      if (switchFromMapToList) {
        this.scrollToListing(selectedListingId)
      } else if (prevProps.uri !== uri) {
        this.listContainerElement.scrollTop = 0
      }
      if (typeof window !== 'undefined' && window.localStorage) {
        const listingId = window.localStorage.getItem('uriListingId')

        if (listingId) {
          window.localStorage.setItem('uriListingId', '')
          this.props.selectListing(listingId)
          this.scrollToListing(listingId)
        }
      }
    }
  }

  componentWillUnmount() {
    if (this.listContainerElement) {
      this.listContainerElement.removeEventListener('scroll', this.lazyCheck, true)
    }
  }

  @autobind
  lazyCheck() {
    forceCheck()
    this.selectListingClosestToTop()
  }

  scrollToListing(listingId, scrollImmediately) {
    const id = `list-${listingId}`
    const node = document.getElementById(id)
    const nodeSort = document.getElementById('sortDropDown')
    const allowScroll = this.lastScrolledId !== id

    if (!node) {
      // reset scroll to top if property card no longer exists
      if (this.listContainerElement) this.listContainerElement.scrollTop = 0
      return
    }

    if (allowScroll) {
      if (scrollImmediately && nodeSort) {
        nodeSort.scrollIntoView()
        this.lastScrolledId = id
      } else {
        setTimeout(() => {
          node.scrollIntoView()
          this.lastScrolledId = id
          if (this.listContainerElement) this.listContainerElement.scrollTop -= 100
        }, 100)
      }
    }
  }

  selectListingClosestToTop() {
    const {
      currentView,
      listings,
      selectListing,
      selectedListingId,
    } = this.props

    if (currentView !== 'list') {
      return
    }

    const htmlId = listings && listings.length ? `list-${listings[0].listingId}` : null

    if (htmlId && this.listContainerElement) {
      const listingCell = document.getElementById(htmlId)

      const positionFromTop = this.listContainerElement ? this.listContainerElement.scrollTop : 0

      const heightOfCell = this.absoluteHeight(listingCell)

      const advertisement = document.getElementById(SMALL_LIST_VIEW_AD_KEY)

      const index = positionFromTop > 0 && heightOfCell > 0
        ? Math.round(positionFromTop / heightOfCell)
        : 0

      let indexWithAdAdjustment

      if (advertisement && index >= AD_POSITION + 1) {
        const heightOfAd = advertisement.offsetHeight
        indexWithAdAdjustment = Math.round((positionFromTop - heightOfAd) / heightOfCell)
      }

      const listing = listings[indexWithAdAdjustment || index]

      const allowSelect = listing && listing.listingId &&
        (selectedListingId !== listing.listingId && selectListing)

      if (allowSelect) {
        selectListing(listing.listingId)
      }
    }
  }

  absoluteHeight(el) {
    if (el) {
      const styles = window.getComputedStyle(el)
      const margin = parseFloat(styles.marginTop) +
        parseFloat(styles.marginBottom)
      return Math.ceil(el.offsetHeight + margin)
    }
    return 0
  }

  render() {
    const {
      city,
      count,
      currentView,
      hidden,
      listings,
      loading,
      loadingOptions,
      nextPageLink,
      openModal,
      pageNumber,
      pathname,
      previousPageLink,
      pushState,
      query,
      selectedListingId,
      selectListing,
      toggleView,
      total,
      locationSlug,
      searchType,
      propertyType,
      refinementCriteria,
      loadedAds,
      renderAd,
      theme,
      title,
    } = this.props

    return (
      <div
        id="ListViewContainer"
        ref={div => { this.listContainerElement = div }}
        hidden={hidden}
        className={theme.List}
        data-tid="list-view"
      >
        <meta
          itemProp="name"
          content={title}
        />
        <BreadCrumbs />
        <Listings
          loadedAds={loadedAds}
          city={city}
          count={count}
          currentView={currentView}
          listings={listings}
          loading={loading}
          loadingOptions={loadingOptions}
          nextPageLink={nextPageLink}
          openModal={openModal}
          pageNumber={pageNumber}
          pathname={pathname}
          previousPageLink={previousPageLink}
          pushState={pushState}
          query={query}
          selectedListingId={selectedListingId}
          selectListing={selectListing}
          toggleView={toggleView}
          total={total}
          viewType="list"
          locationSlug={locationSlug}
          searchType={searchType}
          propertyType={propertyType}
          refinementCriteria={refinementCriteria}
          renderAd={renderAd}
          withPaginationCards
        />
        <SeoContent />
        <Footer withAccordionLinks />
      </div>
    )
  }
}
