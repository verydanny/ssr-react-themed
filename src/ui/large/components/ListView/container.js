import { connect } from 'react-redux'
import {
  selectListing,
  selectListingFromMapPin,
  openModal,
} from 'app/store/shared/actions'
import {
  clearListingData,
  clearListingScroll,
  mapZoomOutOnEmpty,
} from 'app/store/large/actions'
import { toggleFavorite } from 'app/store/shared/actions/user'
import {
  previousPageLink as previousPageLinkSelector,
  nextPageLink as nextPageLinkSelector,
  getPageNumber as pageNumberSelector,
  getCityName,
} from 'app/selectors/criteria'
import {
  getIsLoading as isLoadingSelector,
  selectedListingId as selectedListingIdSelector,
  getSelectedListingOptions,
  getFilteredViewportListings,
  getLoadingOptions,
  getListingScroll,
  getTotal,
} from 'app/selectors/listings'
import {
  getPointsCount,
} from 'app/selectors/map'
import {
  getPageStartIndex,
  getPageEndIndex,
} from 'app/selectors/page'
import {
  query as querySelector,
} from 'app/selectors/router'
import {
  getFavorites,
} from 'app/selectors/user'
import ListView from './ListView'

const mapStateToProps = state => ({
  city: getCityName(state),
  count: getPointsCount(state),
  endIndex: getPageEndIndex(state),
  listings: getFilteredViewportListings(state),
  loading: isLoadingSelector(state),
  loadingOptions: getLoadingOptions(state),
  nextPageLink: nextPageLinkSelector(state),
  pageNumber: pageNumberSelector(state),
  previousPageLink: previousPageLinkSelector(state),
  query: querySelector(state),
  selectedListingId: selectedListingIdSelector(state),
  selectedListingIsClicked: getSelectedListingOptions(state).clicked,
  selectedListingShouldScroll: getListingScroll(state),
  startIndex: getPageStartIndex(state),
  total: getTotal(state),
  favorites: getFavorites(state),
})

const mapDispatchToProps = {
  clearListingData,
  clearListingScroll,
  mapZoomOutOnEmpty,
  openModal,
  selectListing,
  selectListingFromMapPin,
  toggleFavorite,
}

export default connect(mapStateToProps, mapDispatchToProps)(ListView)
