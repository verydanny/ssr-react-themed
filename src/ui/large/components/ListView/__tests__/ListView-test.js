import React from 'react'
import { mount, shallow } from 'enzyme'
import { createMockStore } from 'mocks'
import { initialState as criteria } from 'app/store/reducers/criteria'
import { CookiesProvider } from 'react-cookie'
import { Provider } from 'react-redux'
import ListView from '../ListView'

const ListViewUnwrapped = ListView.WrappedComponent

describe('ListView', () => {
  let component
  let store
  const state = {
    criteria: {
      ...criteria,
    },
    meta: {},
  }

  const defaultProps = {
    theme: {},
    listings: [],
    openModal: () => {},
    total: 0,
    selectListingFromMapPin: () => {},
    clearListingData: () => {},
    loading: false,
    loadingOptions: {},
    clearListingScroll: jest.fn(),
  }

  const setup = propsOverride => {
    const props = { ...defaultProps, ...propsOverride }
    store = createMockStore(state)
    component = mount(
      <CookiesProvider>
        <Provider store={store}>
          <ListView {...props} />
        </Provider>
      </CookiesProvider>
    )
    return component
  }

  beforeEach(() => setup())

  it('renders zero result card', () => {
    const wrapper = setup()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders listings', () => {
    const wrapper = setup({ total: 10 })
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('scroll functionality', () => {
    describe('selectedListingShouldScroll is set', () => {
      const fnScroll = jest.fn()
      const fnClear = jest.fn()
      ListViewUnwrapped.prototype.scrollToSelectedListing = fnScroll
      setup({ selectedListingShouldScroll: true, clearListingScroll: fnClear })

      it('calls scrollToSelectedListing', () => {
        expect(fnScroll).toBeCalled()
      })
      it('calls clearListingScroll', () => {
        expect(fnClear).toBeCalled()
      })
    })

    describe('selectedListingShouldScroll is not set', () => {
      const fnScroll = jest.fn()
      ListViewUnwrapped.prototype.scrollToSelectedListing = fnScroll
      setup({ selectedListingShouldScroll: false })

      it('does not call scrollToSelectedListing', () => {
        expect(fnScroll).not.toBeCalled()
      })
    })
  })

  describe('hover functionality', () => {
    describe('handleCardMouseEnter is triggered', () => {
      const fnSelectMapPin = jest.fn()
      const props = {
        ...defaultProps,
        selectListingFromMapPin: fnSelectMapPin,
      }
      const wrapper = shallow(<ListViewUnwrapped {...props} />)
      const instance = wrapper.instance()
      instance.handleCardMouseEnter({}, 'foo')

      it('calls selectListingFromMapPin', () => {
        expect(fnSelectMapPin).toBeCalledWith('foo', {
          clicked: true,
          trackClick: false,
        })
      })
    })
  })
})
