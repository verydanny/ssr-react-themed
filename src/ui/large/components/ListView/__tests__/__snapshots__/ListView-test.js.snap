// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListView renders listings 1`] = `

<div>
  <div>
    <div class>
      <form action="?"
            data-tag_section="search_input"
      >
        <input data-tid="search-location"
               type="search"
               name="searchInput"
               aria-label="Search"
               class
               placeholder="City, State or Zip"
               data-tag_item="search_bar"
               autocomplete="off"
               autocorrect="off"
               value
        >
      </form>
    </div>
    <div data-tag_item="filters_max_price"
         data-tag_section="quick_filters"
    >
      <div class>
        <button data-tid="dropdown-anchor"
                themed="[object Object]"
                class
        >
          <span>
          </span>
          <span>
            Max Price
          </span>
        </button>
      </div>
    </div>
    <div data-tag_item="filters_bedrooms"
         data-tag_section="quick_filters"
    >
      <div class>
        <button data-tid="dropdown-anchor"
                themed="[object Object]"
                class
        >
          <span>
          </span>
          <span>
            Beds
          </span>
        </button>
      </div>
    </div>
    <div class
         data-tag_item="filters"
         data-tag_section="more_filters"
         role="presentation"
         data-tid="more-filters"
    >
      <div>
        More filters
      </div>
    </div>
  </div>
  <ul data-tag_section="filter_tokens"
      data-tag_item="filter_tokens"
      data-tid="filter-tokens"
  >
  </ul>
  <div data-tid="list-view">
    <div>
      <div>
        <div data-tid="property-count"
             class
        >
          10 Properties
        </div>
        <div class
             data-tid="sort-filter"
             data-tag_item="sort_dropdown"
        >
          <button data-tid="dropdown-anchor"
                  themed="[object Object]"
                  class
          >
            <span>
              Sort by:
            </span>
            <span>
              Best Match
            </span>
          </button>
        </div>
      </div>
      <div data-tag_section="property_listings">
      </div>
      <div>
        <div data-tid="seo-content">
          <div>
          </div>
          <div class
               data-tid="srp-bottom-ad"
               id="gpt_list_srp_desktop_bottom"
          >
            <div class="adunitContainer">
              <div id="adSlot-5"
                   class="adBox"
              >
              </div>
            </div>
          </div>
        </div>
        <div class
             data-tid="srp-bottom-ad2"
             id="gpt_list_srp_desktop_bottom_ad2"
        >
          <div class="adunitContainer">
            <div id="adSlot-6"
                 class="adBox"
            >
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <footer data-tag_section="footer">
        <div>
          <div data-tid="rent-logo-footer">
            <a href="/">
            </a>
          </div>
          <div>
            <div data-tid="our-company"
                 class
            >
              <h4>
                <button>
                  Our Company
                </button>
              </h4>
              <ul>
                <li>
                  <a href="/company/company-overview/"
                     data-tag_item="about_us"
                  >
                    About Us
                  </a>
                </li>
                <li>
                  <a href="/company/press/contacts/"
                     data-tag_item="media_contact"
                  >
                    Press &amp; Media
                  </a>
                </li>
                <li>
                  <a href="/careers/jobs/"
                     data-tag_item="careers"
                  >
                    Careers
                  </a>
                </li>
                <li>
                  <a href="/company/privacy-full/"
                     data-tag_item="privacy_policy"
                  >
                    Privacy Policy
                  </a>
                </li>
                <li>
                  <a href="/company/legal/termsofservice/"
                     data-tag_item="terms_of_service"
                  >
                    Terms of Service
                  </a>
                </li>
              </ul>
            </div>
            <div data-tid="popular-searches"
                 class
            >
              <h4>
                <button>
                  Popular Searches
                </button>
              </h4>
              <ul>
                <li>
                  <a href="/apartments-by-city"
                     data-tag_item="popular_city"
                  >
                    Popular Cities
                  </a>
                </li>
                <li>
                  <a href="/search-by-state"
                     data-tag_item="us_state"
                  >
                    Search by State
                  </a>
                </li>
                <li>
                  <a href="/rentals-by-type"
                     data-tag_item="rental_type"
                  >
                    Rental Type
                  </a>
                </li>
                <li>
                  <a href="/off-campus-housing"
                     data-tag_item="college_university"
                  >
                    College &amp; University
                  </a>
                </li>
                <li>
                  <a href="/housing-near-military-base"
                     data-tag_item="military"
                  >
                    Military Base
                  </a>
                </li>
              </ul>
            </div>
            <div data-tid="let-us-help"
                 class
            >
              <h4>
                <button>
                  Let Us Help
                </button>
              </h4>
              <ul>
                <li>
                  <a href="/company/renter-help/"
                     data-tag_item="renter_help"
                  >
                    Help for Renters
                  </a>
                </li>
                <li>
                  <a href="/company/security/"
                     data-tag_item="rental_safety"
                  >
                    Avoid Scams
                  </a>
                </li>
                <li>
                  <a href="/manage"
                     data-tag_item="list_a_property"
                  >
                    List a Property
                  </a>
                </li>
                <li>
                  <a href="/company/renter-help/"
                     data-tag_item="contact_support"
                  >
                    Contact Support
                  </a>
                </li>
                <li>
                  <a href="/sitemap"
                     data-tag_item="sitemap"
                  >
                    Site Map
                  </a>
                </li>
              </ul>
            </div>
          </div>
          <div>
            <div style="height: 30px;"
                 class="lazyload-placeholder"
            >
            </div>
          </div>
          <div data-tid="footer-app-links">
            Download the Rent.com app for
            <a href="https://play.google.com/store/apps/details?id=com.rent&amp;referrer=ts%3D96e725f6-e305-11e4-9781-00163cabc4e4"
               target="_blank"
               rel="external noopener noreferrer"
               data-tag_item="android_download_button"
            >
              Android
            </a>
            and
            <a href="https://itunes.apple.com/us/app/rent.com-apartments-home-rentals/id388038507?mt=8"
               target="_blank"
               rel="external noopener noreferrer"
               data-tag_item="ios_download_button"
            >
              iOS
            </a>
            .
          </div>
        </div>
        <div>
          <div>
            <div data-tid="rentpath-copyright-text">
              Â© 2018 RentPath, LLC.  All rights reserved.
              <a href="/company/legal/termsofservice/"
                 data-tag_item="terms_of_service_bottom"
              >
                Terms of Service
              </a>
              and
              <a href="/company/privacy-full/"
                 data-tag_item="privacy_policy_bottom"
              >
                Privacy Policy
              </a>
              .
            </div>
            <a href="/company/legal/equal/"
               data-tag_item="equal_housing"
               title="Equal Housing Opportunity"
            >
            </a>
          </div>
        </div>
      </footer>
      <div class
           data-tid="srp-footer-ad"
           id="gpt_list_srp_footer_ad"
      >
        <div class="adunitContainer">
          <div id="adSlot-7"
               class="adBox"
          >
          </div>
        </div>
      </div>
    </div>
  </div>
</div>

`;

exports[`ListView renders zero result card 1`] = `

<div>
  <div>
    <div class>
      <form action="?"
            data-tag_section="search_input"
      >
        <input data-tid="search-location"
               type="search"
               name="searchInput"
               aria-label="Search"
               class
               placeholder="City, State or Zip"
               data-tag_item="search_bar"
               autocomplete="off"
               autocorrect="off"
               value
        >
      </form>
    </div>
    <div data-tag_item="filters_max_price"
         data-tag_section="quick_filters"
    >
      <div class>
        <button data-tid="dropdown-anchor"
                themed="[object Object]"
                class
        >
          <span>
          </span>
          <span>
            Max Price
          </span>
        </button>
      </div>
    </div>
    <div data-tag_item="filters_bedrooms"
         data-tag_section="quick_filters"
    >
      <div class>
        <button data-tid="dropdown-anchor"
                themed="[object Object]"
                class
        >
          <span>
          </span>
          <span>
            Beds
          </span>
        </button>
      </div>
    </div>
    <div class
         data-tag_item="filters"
         data-tag_section="more_filters"
         role="presentation"
         data-tid="more-filters"
    >
      <div>
        More filters
      </div>
    </div>
  </div>
  <ul data-tag_section="filter_tokens"
      data-tag_item="filter_tokens"
      data-tid="filter-tokens"
  >
  </ul>
  <div data-tid="list-view">
    <div class
         data-tid="zeroResultsCard"
    >
      <div>
        <div data-tid="zero_properties">
          0 Properties Found
        </div>
        <div data-tid="suggest_text">
          Try adjusting the filters above or zoom out on the map
        </div>
      </div>
      <div>
        <div data-tid="next_steps">
          Next Steps :
        </div>
        <button class
                data-tid="adjust_filters"
        >
          Adjust Filters
        </button>
      </div>
      <div>
        <span>
          or
        </span>
        <button data-tid="zoom_out">
          Zoom Out
        </button>
      </div>
    </div>
    <div>
      <footer data-tag_section="footer">
        <div>
          <div data-tid="rent-logo-footer">
            <a href="/">
            </a>
          </div>
          <div>
            <div data-tid="our-company"
                 class
            >
              <h4>
                <button>
                  Our Company
                </button>
              </h4>
              <ul>
                <li>
                  <a href="/company/company-overview/"
                     data-tag_item="about_us"
                  >
                    About Us
                  </a>
                </li>
                <li>
                  <a href="/company/press/contacts/"
                     data-tag_item="media_contact"
                  >
                    Press &amp; Media
                  </a>
                </li>
                <li>
                  <a href="/careers/jobs/"
                     data-tag_item="careers"
                  >
                    Careers
                  </a>
                </li>
                <li>
                  <a href="/company/privacy-full/"
                     data-tag_item="privacy_policy"
                  >
                    Privacy Policy
                  </a>
                </li>
                <li>
                  <a href="/company/legal/termsofservice/"
                     data-tag_item="terms_of_service"
                  >
                    Terms of Service
                  </a>
                </li>
              </ul>
            </div>
            <div data-tid="popular-searches"
                 class
            >
              <h4>
                <button>
                  Popular Searches
                </button>
              </h4>
              <ul>
                <li>
                  <a href="/apartments-by-city"
                     data-tag_item="popular_city"
                  >
                    Popular Cities
                  </a>
                </li>
                <li>
                  <a href="/search-by-state"
                     data-tag_item="us_state"
                  >
                    Search by State
                  </a>
                </li>
                <li>
                  <a href="/rentals-by-type"
                     data-tag_item="rental_type"
                  >
                    Rental Type
                  </a>
                </li>
                <li>
                  <a href="/off-campus-housing"
                     data-tag_item="college_university"
                  >
                    College &amp; University
                  </a>
                </li>
                <li>
                  <a href="/housing-near-military-base"
                     data-tag_item="military"
                  >
                    Military Base
                  </a>
                </li>
              </ul>
            </div>
            <div data-tid="let-us-help"
                 class
            >
              <h4>
                <button>
                  Let Us Help
                </button>
              </h4>
              <ul>
                <li>
                  <a href="/company/renter-help/"
                     data-tag_item="renter_help"
                  >
                    Help for Renters
                  </a>
                </li>
                <li>
                  <a href="/company/security/"
                     data-tag_item="rental_safety"
                  >
                    Avoid Scams
                  </a>
                </li>
                <li>
                  <a href="/manage"
                     data-tag_item="list_a_property"
                  >
                    List a Property
                  </a>
                </li>
                <li>
                  <a href="/company/renter-help/"
                     data-tag_item="contact_support"
                  >
                    Contact Support
                  </a>
                </li>
                <li>
                  <a href="/sitemap"
                     data-tag_item="sitemap"
                  >
                    Site Map
                  </a>
                </li>
              </ul>
            </div>
          </div>
          <div>
            <div style="height: 30px;"
                 class="lazyload-placeholder"
            >
            </div>
          </div>
          <div data-tid="footer-app-links">
            Download the Rent.com app for
            <a href="https://play.google.com/store/apps/details?id=com.rent&amp;referrer=ts%3D96e725f6-e305-11e4-9781-00163cabc4e4"
               target="_blank"
               rel="external noopener noreferrer"
               data-tag_item="android_download_button"
            >
              Android
            </a>
            and
            <a href="https://itunes.apple.com/us/app/rent.com-apartments-home-rentals/id388038507?mt=8"
               target="_blank"
               rel="external noopener noreferrer"
               data-tag_item="ios_download_button"
            >
              iOS
            </a>
            .
          </div>
        </div>
        <div>
          <div>
            <div data-tid="rentpath-copyright-text">
              Â© 2018 RentPath, LLC.  All rights reserved.
              <a href="/company/legal/termsofservice/"
                 data-tag_item="terms_of_service_bottom"
              >
                Terms of Service
              </a>
              and
              <a href="/company/privacy-full/"
                 data-tag_item="privacy_policy_bottom"
              >
                Privacy Policy
              </a>
              .
            </div>
            <a href="/company/legal/equal/"
               data-tag_item="equal_housing"
               title="Equal Housing Opportunity"
            >
            </a>
          </div>
        </div>
      </footer>
      <div class
           data-tid="srp-footer-ad"
           id="gpt_list_srp_footer_ad"
      >
        <div class="adunitContainer">
          <div id="adSlot-3"
               class="adBox"
          >
          </div>
        </div>
      </div>
    </div>
  </div>
</div>

`;
